software/bsg_manycore_lib/bsg_tile_group_barrier.hpp

```diff
@@ -0,0 +1,285 @@
+        // participating, there is nothing to be done
+        if (BARRIER_Y_DIM == 1 && BARRIER_X_DIM == 1)
+                return;
+
+        int center_x_cord = (this->r_barrier._x_cord_start + this->r_barrier._x_cord_end) / 2;
+        int center_y_cord = (this->c_barrier._y_cord_start + this->c_barrier._y_cord_end) / 2;
+
+        #ifdef BSG_BARRIER_DEBUG
+                if( bsg_x == center_x_cord && bsg_y == center_y_cord ){
+                        bsg_print_time();
+                }
+        #endif
+
+        // Send sync signals to the center tile of each row 
+        r_barrier.sync(center_x_cord);
+
+        // Wait on sync signals from all tiles in the row to be received
+        // Send sync signals to the center tile of the center col
+        // Only performed by tiles in the center column
+        if( bsg_x == center_x_cord) {
+                this->r_barrier.wait_on_sync();
+                this->c_barrier.sync(center_x_cord, center_y_cord);
+        }
+
+        // Send alert to all tiles of the col
+        // Wait on sync signals from all tiles in the center column to be received
+        // Send alert signals from the center tile of the center column to all
+        // tiles in the cetner column
+        // Reset the local done list of the bsg_col_barrier class
+        // Only performed by the center tile of the center column
+        if( bsg_x == center_x_cord && bsg_y == center_y_cord) { 
+                this->c_barrier.wait_on_sync();
+                this->c_barrier.reset();
+                this->c_barrier.alert();
+        }
+
+        // Wait on alert signals from center tile of the center column to be received
+        // Send alert signals from center tile of the row to all tiles in the row
+        // Reset the local done list of the bsg_row_barrier class
+        // Performed by all tiles in the center column 
+        if( bsg_x == center_x_cord) {
+                this->c_barrier.wait_on_alert();
+                this->r_barrier.reset();
+                this->r_barrier.alert();
+        }
+
+        // Wait on alert signal from the center tile in each row 
+        // Once the alert signal is received, all tiles are syncrhonized
+        // and can carry on
+        this->r_barrier.wait_on_alert();
+
+        #ifdef BSG_BARRIER_DEBUG
+                if( bsg_x == center_x_cord && bsg_y == center_y_cord ){
+                        bsg_print_time();
+                }
+        #endif
+        return;
+    };
+};
+
+
+// Helper funcitons.
+// check if the char array are all non-zeros
+inline void poll_range( int range, volatile const unsigned int *p){
+        int i;
+        do{
+                for( i= 0; i <= range; i++) {
+                        if ( p[ i ] == 0) break;
+                }
+        }while ( i <= range);
+}
+
+// Wait until the specified memory address is written by other tiles
+// with specific value
+inline int bsg_wait_local_int(int * ptr,  int cond ) {
```

Comment:
Isn't this function already in bsg_mutex.h? I'm not sure why it needs to be duplicated here.

Reply:
I'm not necessary suggesting that this file should include "bsg_mutex.h"; there might be a better way to organize shared codes.

Reply:
Oh you're right my bad. Fixed, thanks!